src/jit/lsraxarch.cpp

```diff
@@ -1013,7 +1013,21 @@ void LinearScan::TreeNodeInfoInitShiftRotate(GenTree* tree)
     if (!shiftBy->isContained())
     {
         shiftBy->gtLsraInfo.setSrcCandidates(this, RBM_RCX);
-        if (!source->isContained())
+        if (source->isContained())
+        {
+            if (source->OperIs(GT_IND))
+            {
+                if (source->AsIndir()->Base() != nullptr)
+                {
+                    source->AsIndir()->Base()->gtLsraInfo.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RCX);
+                }
+                if (source->AsIndir()->Index() != nullptr)
+                {
+                    source->AsIndir()->Index()->gtLsraInfo.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RCX);
+                }
+            }
+        }
+        else
         {
             source->gtLsraInfo.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RCX);
```

Comment:
Do we need to do this because `shiftBy` is fixed to be `rcx`?

Reply:
This kind of shift  instruction has a contained read/modify/write op1:

shl  [rbx+rdx*8+8], 3

Reply:
Sergey  came by and we chatted and we are not sure why you need to exclude RCX here.

I particular I am worried that we could have  a case like this:

   a[i] <<= i;

Where the index register would like to be in RCX.

Reply:
If we did not exclude RCX, it would still be correct, as you say. If a source was in RCX, it would be copied or spilled. Unfortunately, it is more often spilled than copied (more work is needed to use copies in cases like these). However, making this change is a net positive, because it is a stronger preventative for conflicts that require a copy.
If it is a lclVar (e.g. the case you describe), doing this will remove RCX from the preference set of 'i', but won't prevent it from getting that register, e.g. if it has a fixed requirement for that register elsewhere.